File: src/engine/adapters/redis/caps.ts
import type { Caps } from '../../../shared/caps';

// §5.1's redis row: key/value-shaped, cursor (SCAN) pagination, no FK navigation, a shell-style
// console (§8.14). Writable as of this phase — string-only edit/insert plus type-agnostic delete
// (see the canInsert/canUpdate/canDelete comment below).
export const redisCaps: Caps = {
  tabular: false,
  documents: false,
  keyValue: true,
  stream: false,
  defaultPageKind: 'keyvalue',
  sql: true,
  // P23 D10: stays false permanently, not deferred — a key's type/TTL/memory usage are already on
  // every KeyValuePage and already rendered by KeyValueView.vue (redis/read.ts). A definition tab
  // would be a second, staler view of the same three facts for a node whose only other property is
  // its name.
  definition: false,
  projection: false,
  serverFilter: false,
  // Per-key counts use O(1) exact type-length commands (HLEN/SCARD/ZCARD/LLEN/XLEN, or 1 for a
  // string) — D6. Distinct from §5.1's "DBSIZE only" wording, which describes a tree/db-wide
  // count this adapter never surfaces through count().
  exactCount: true,
  pagination: 'cursor',
  foreignKeys: false,
  // index.ts's mutate() now backs all three: `insert` (SET ... NX, a brand-new key, string-typed
  // only), `update` (a plain SET, also string-typed only — mutate.ts's assertEditableType), and
  // `delete` (DEL, type-agnostic). Editing/creating a hash/list/set/zset/stream element needs its
  // own per-type semantics (HSET a field, LSET an index, SADD/SREM, ZADD, XADD) — a materially
  // bigger job, out of scope for this version; the UI disables Edit/Add for those cases.
  canInsert: true,
  canUpdate: true,
  canDelete: true,
  writable: true,
  transactions: false,
  cancel: true, // ctx.signal between bounded SCAN-family rounds is fully effective (D7/D8)
};

File: apps/kira-studio/internal/adapters/redis/caps.go
package redis

import (
	"github.com/kirathecat/kira-studio/apps/kira-studio/internal/adapters"
	"github.com/kirathecat/kira-studio/apps/kira-studio/internal/page"
)

// caps is caps.ts's redisCaps, literally (§4.3, C18). Key/value-shaped, cursor (SCAN) pagination,
// no FK navigation, a shell-style console. Cancel is true and honest despite Cancel() returning
// false permanently (C9) — CheckCancelled between bounded SCAN-family rounds is fully effective.
var caps = adapters.Caps{
	Tabular:         false,
	Documents:       false,
	KeyValue:        true,
	Stream:          false,
	KeyBrowser:      true, // P41: a db index's key namespace is unbounded — browsed in a Browse tab
	DefaultPageKind: page.PageKindKeyValue,
	SQL:             true,
	// P23 D10: stays false permanently — a key's type/TTL/memory usage are already on every
	// KeyValuePage.
	Definition:   false,
	Describe:     false,
	Projection:   false,
	ServerFilter: false,
	// Per-key counts use O(1) exact type-length commands (HLEN/SCARD/ZCARD/LLEN/XLEN, or 1 for a
	// string).
	ExactCount:  true,
	Pagination:  adapters.PaginationCursor,
	ForeignKeys: false,
	// mutate.go backs all three: insert (SET ... NX, string-typed only), update (a plain SET,
	// also string-typed only), delete (DEL, type-agnostic).
	CanInsert:    true,
	CanUpdate:    true,
	CanDelete:    true,
	Writable:     true,
	Transactions: false,
	Cancel:       true,
	FileTransfer: false,
}
